Mathlib/Algebra/MonoidAlgebra/Degree.lean

```diff
@@ -222,35 +223,301 @@ variable [Semiring R]
 
 section SupDegree
 
-variable [AddZeroClass A] [SemilatticeSup B] [AddZeroClass B] [OrderBot B]
-  (D : A → B)
+variable [AddZeroClass A] [Add B] (D : A → B)
+
+section SemilatticeSup
+
+variable [SemilatticeSup B] [OrderBot B]
 
 /-- Let `R` be a semiring, let `A, B` be two `AddZeroClass`es, let `B` be an `OrderBot`,
 and let `D : A → B` be a "degree" function.
 For an element `f : R[A]`, the `R[A]`, the element `supDegree f : B` is the
 supremum of all the elements in the support of `f`, or `⊥` if `f` is zero.
 Often, the Type `B` is `WithBot A`,
 If, further, `A` has a linear order, then this notion coincides with the usual one,
-using the maximum of the exponents. -/
-@[reducible]
-def supDegree (f : R[A]) : B :=
-  f.support.sup D
+using the maximum of the exponents.
+
+If `A := σ →₀ ℕ` then `R[A] = MvPolynomial σ R`, and if we equip `σ` with a linear order then
+the induced linear order on `Lex A` equips `MvPolynomial` ring with a
+[monomial order](https://en.wikipedia.org/wiki/Monomial_order) (i.e. a linear order on `A`, the
+type of (monic) monomials in `R[A]`, that respects addition). We make use of this monomial order
+by taking `D := toLex`, and different monomial orders could be accessed via different type
+synonyms once they are added. -/
+@[reducible] def supDegree (f : R[A]) : B := f.support.sup D
 
-theorem supDegree_add_le (f g : R[A]) :
+/-- If `D` is an injection into a linear order `B`, the leading coefficient of `f : R[A]` is the
+  nonzero coefficient of highest degree according to `D`, or 0 if `f = 0`. In general, it is defined
+  to be the coefficient at an inverse image of `supDegree f` (if such exists). -/
+noncomputable def leadingCoeff (f : R[A]) : R := f (D.invFun <| f.supDegree D)
+
+/-- An element `f : R[A]` is monic if its leading coefficient is one. -/
+@[reducible] def Monic (f : R[A]) : Prop := f.leadingCoeff D = 1
+
+variable {D}
+
+theorem supDegree_add_le {f g : R[A]} :
     (f + g).supDegree D ≤ (f.supDegree D) ⊔ (g.supDegree D) :=
   sup_support_add_le D f g
```

Comment:
Note that this one is first proved without the `supDegree` wrapper in a previous file; arguably the `neg`/`sub`/`sum` lemmas should be handled in the same way.

Reply:
It's proved in the same file. I think I can just remove `sup_support_add_le`.

Reply:
Hmm @adomani what do you think of moving `sup/infDegree` to the top of the file and replace all `.support.sup` by `supDegree`?

Reply:
I like this idea!

This file grew out of an attempt at defining *as late as possible* the new "degrees", just because their definition was still in discussion.  Now that the definition is "accepted", I'm all in favour of using it as soon as we can!